### PR TITLE
Process reparse points to Microsoft Store applications

### LINKED ISSUE #9970: Windows Store applications incorrectly assumed to be console applications
## Steps to reproduce

Run a Windows program installed in `${env:USERPROFILE}\AppData\Local\Microsoft\WindowsApps`, e.g. `iTunes.exe` or `wt.exe` (the new Windows Terminal) from the interactive command line.

## Expected behavior

The process launches in a new window, a new prompt is displayed and you can run new commands in PowerShell.

## Actual behavior

PowerShell waits until the Windows application exits before allowing new commands to be entered. PowerShell is behaving as though the program was a console application.

It makes perfect sense to wait when PowerShell can't determine if the process is a Windows or console application, so a proper should do a better job processing these reparse points to inspect the real executable.

### PR DESCRIPTION
**Pull request by iSazonov**
# PR Summary

Fix #9970

- Add resolving a Store application path (that is a reparse point) to a real path.
        SHGetFileInfo() does not understand reparse points and returns 0 ("non exe or error") so we are trying to get a real path before.

- Exclude Windows code from compilation on Unix.
- Exclude second extra check if an app is Windows application or console application
    Add new properties (IsWindowsApplication/IsConsoleApplication) to NativeCommandProcessor to cache the check.
    Before the PR we did the check twice - (1) in redirection check, (2) in background check.

I tested this manually with `C:\Users\<username>\AppData\Local\Microsoft\WindowsApps\Skype.exe`.

/cc @DHowett

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary. This may include:
        - Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode (which runs in a different PS Host).
        - Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
        - Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
        - Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).

### CODE REVIEW

**Comment by rjmholt** [comment]
I'm afraid I don't know much about reparse points or the Microsoft Store. Requesting review from @SteveL-MSFT, @anmenaga and @daxian-dbw.

**Review by SteveL-MSFT — APPROVED** [review]

**Comment by DHowett** [comment]
This feels like the right solution to me, especially because we're already treating IO_REPARSE_TAG_APPEXECLINK as though it were a link already.

**Comment by rjmholt** [comment]
CI failures seem to be unrelated -- will see if there's a fix

File: src/System.Management.Automation/namespaces/FileSystemProvider.cs [hunk 1]
```diff
@@ -8380,6 +8371,21 @@ internal static bool WinIsHardLink(ref IntPtr handle)
...
+        {
+            // We set accessMode parameter to zero because documentation says:
+            // If this parameter is zero, the application can query certain metadata
+            // such as file, directory, or device attributes without accessing
+            // that file or device, even if GENERIC_READ access would have been denied.
+            using (SafeFileHandle handle = OpenReparsePoint(path, FileDesiredAccess.GenericZero))
+            {
+                string linkTarget = WinInternalGetTarget(handle);
+
+                return linkTarget;
```
Comment (rjmholt) [inline src/System.Management.Automation/namespaces/FileSystemProvider.cs:8385]:
```suggestion
                return WinInternalGetTarget(handle);
```

**Comment by iSazonov** [comment]
@rjmholt I rebased to pass CIs.

**Pull request merged**